README.md
A little Hangman-Game made in Python. I'm planning to give the game a GUI!


modules/changelog.txt
19.09.2022
-----------

-Added following commands:
    -!view_words
    -!add_word
    -!sort_words

-Updated "!fix" commands:
    !fix should now correctly reset all variables back to default!

-Fixed a bug that deleted a word when adding another
-Fixed a bug where "!view_word" broke when adding another word
-Fixed a bug that display the wrong amount if lives left

01.10.2022
-----------
-If game is played via a .exe, the console should clear after every guess!



modules/draw.py
from colorama import *

red = Fore.RED
green = Fore.GREEN
blue = Fore.BLUE
yellow = Fore.YELLOW


class Hangman():
    def fail1(self, lives):
        print("=" * 25)
        print("___|___\n")
        print(red + "You have " + yellow + lives + " lives" + red + " remaining!")
        print("=" * 25)

    def fail2(self, lives):
        print("=" * 25)
        print("    _____ \n"
              "   |      \n"
              "   |      \n"
              "   |      \n"
              "   |      \n"
              "   |      \n"
              "   |      \n"
              "___|___\n")
        print(red + "You have " + yellow + lives + " lives" + red + " remaining!")
        print("=" * 25)

    def fail3(self, lives):
        print("=" * 25)
        print("    _____ \n"
              "   |     |\n"
              "   |     |\n"
              "   |      \n"
              "   |      \n"
              "   |      \n"
              "   |      \n"
              "___|___\n")
        print(red + "You have " + yellow + lives + " lives" + red + " remaining!")
        print("=" * 25)

    def fail4(self, lives):
        print("    _____ \n"
              "   |     |\n"
              "   |     |\n"
              "   |     O\n"
              "   |      \n"
              "   |      \n"
              "   |      \n"
              "___|___\n")
        print(red + "You have " + yellow + lives + " lives" + red + " remaining!")
        print("=" * 25)

    def fail5(self, lives):
        print("    _____   \n"
              "   |     |  \n"
              "   |     |  \n"
              "   |     O  \n"
              "   |    /|\ \n"
              "   |        \n"
              "   |        \n"
              "___|___\n")
        print(red + "You have " + yellow + lives + " lives" + red + " remaining!")
        print("=" * 25)

    def fail6(self, lives):
        print("    _____   \n"
              "   |     |  \n"
              "   |     |  \n"
              "   |     O  \n"
              "   |    /|\ \n"
              "   |     |  \n"
              "   |    / \ \n"
              "___|___\n")
        print(red + "You have " + yellow + lives + " lives" + red + " remaining!")
        print("=" * 25)

    def img_print(self, fails, lives):
        if fails == 1:
            hm.fail1(lives)
            return
        elif fails == 2:
            hm.fail2(lives)
            return
        elif fails == 3:
            hm.fail3(lives)
            return
        elif fails == 4:
            hm.fail4(lives)
            return
        elif fails == 5:
            hm.fail5(lives)
            return
        elif fails == 6:
            hm.fail6(lives)
            return
        else:
            return


hm = Hangman()


modules/commands.py
import time
import random
from colorama import *
from inspect import getmembers, isfunction

from main import reset

# define colors
red = Fore.RED
green = Fore.GREEN
blue = Fore.BLUE
yellow = Fore.YELLOW

version = "4.3.2"


def commands_main():
    x = []
    for i in getmembers(Commands, isfunction):
        x.append("!" + i[0])
    print("-" * 50)
    print(yellow + "The commands are:")
    [print(blue + "- " + i) for i in x]
    print(blue + "- !help (to show this again!)")
    print("-" * 50)

    def command_select():
        y = ""
        while y not in x and y != "e" and y != "!help":
            y = input("What command do you want to use? (\"e\" to exit): ")

        if y == "e":
            print("-" * 50)
            return

        elif y == "!help":
            commands_main()

        elif y == x[0]:
            cmds.add_word()
            command_select()

        elif y == x[1]:
            cmds.changelog()
            command_select()

        elif y == x[2]:
            cmds.fix()
            command_select()

        elif y == x[3]:
            cmds.info()
            command_select()

        elif y == x[4]:
            cmds.sort_words()
            command_select()

        elif y == x[5]:
            cmds.view_words()
            command_select()
        else:
            return

    command_select()


class Commands():
    def add_word(self):
        x = None
        while x is None or len(x) < 3 or len(x) > 10 or not x.isalpha():
            print("-" * 30)
            x = input("Input the word you want to add (3-10 characters): ")
            if not x.isalpha():
                print("-" * 30)
                print(red + "Only use letters!")
            elif len(x) < 3:
                print("-" * 30)
                print(red + "Your word must be" + yellow + " at least 3" + red + " characters long!")
            elif len(x) > 10:
                print("-" * 30)
                print(red + "Your word" + yellow + "  can't be longer than 10 " + red + "characters!")

        print("-" * 30)
        y = input("Are you sure you want to add \"" + x + "\" to the word-list? (y/n): ")
        if y.lower() == "n":
            self.add_word()
        if y.lower() == "y":
            with open("words.txt", "a") as file:
                file.writelines(x.lower() + "\n")
        print("-" * 30)
        print(green + "Successfully added your word to the list!")
        print("-" * 30)

    def changelog(self):
        print("-" * 50)

        with open("modules/changelog.txt") as data:
            cl = data.read()
            print(blue + "Changelog for version " + version + "\n\n" + cl)
        print("-" * 50)

    def fix(self):
        print("-" * 50)
        print(red + "The program tries to fix the error, please wait a few seconds!")

        time.sleep(random.random()*4)
        print(blue + "Done!")
        reset()
        print("-" * 50)

    def info(self):
        print("-" * 50)
        print(blue + "Version: " + yellow + version)
        print(blue + "Developer: Jackstar#2002")
        print(blue + "Lines of code: 603")
        print(blue + "Hours of development: 26.5")
        print("-" * 50)

    def sort_words(self):
        print("-" * 30)
        i = ""
        while i.lower() not in ["y", "n"]:
            i = input("Are you sure you want to sort the list of words? (y/n): ")
        if i == "n":
            print("-" * 30)
            return

        else:
            list = []
            with open("words.txt.", "r") as data:
                for x in data.readlines():
                    list.append(x.lower())

            list.sort()
            with open("words.txt", "w") as file:
                for x in list:
                    file.write(x)
        print(green + "Successfully sorted your list of words!")
        print("-" * 30)

    def view_words(self):
        with open("words.txt", "r") as data:
            for idx, i in enumerate(data.readlines()):
                print(blue + str(idx + 1) + ". " + i, end="")
        print()
        print("-" * 30)


cmds = Commands()


main.py
import random
import os
import modules.commands as commands

from modules.draw import Hangman
from colorama import *

hm = Hangman()
init(autoreset=True)

# colors
red = Fore.RED
green = Fore.GREEN
blue = Fore.BLUE
yellow = Fore.YELLOW

# variable definitions
words = []
guess = str()
current_word = []
already_guessed = []
word = str()
lives = 6
fails = 0
got_hint1 = False
hint = None
possible_hints = []


def generate_word():
    global word
    i = 0
    with open("words.txt", "r") as f:
        for line in f.readlines():
            words.append(line)
    for _ in words:
        words[i] = words[i].strip()
        i += 1

    word = str(random.choice(words)).lower()
    return word


def reset():
    global fails
    global word
    global words
    global lives
    global already_guessed
    global guess
    global current_word
    global hint
    word = str()
    words = []
    lives = 6
    already_guessed = []
    guess = ""
    current_word = []
    hint = None
    fails = 0

    input("Press any key to restart... ")
    os.system('cls')


def empty_word():
    global current_word
    for i in range(0, len(word)):
        current_word.append("_ ")

    s = ""
    for x in current_word:
        s += x

    print(blue + "Your word is: " + yellow + s)
    print("-" * 50)


def get_hint():
    global got_hint1
    global hint
    global current_word
    global possible_hints

    hint_no = random.randint(1, 2)

    print("-" * 30)
    print(blue + "You got a hint!")

    if hint_no == 1:
        ind = 0
        for i in current_word:
            if i == "_ ":
                possible_hints += list(word[ind])
            ind += 1
        if len(possible_hints) <= 1:
            print(red + "You are only missing one letter, not hint for you!")
            print("-" * 30)
            return
        print(blue + "A letter got guessed and placed for you!")

        hint = random.choice(possible_hints)
        inx = []
        for idx, value in enumerate(word):
            if value == hint:
                inx.append(idx)

        for i in inx:
            current_word[i] = hint + " "

        got_hint1 = True

    elif hint_no == 2:
        vowel_count = 0
        for x in list(word):
            if x in ["a", "e", "i", "o", "u"]:
                vowel_count += 1
        print(blue + "Your word has: " + yellow + str(vowel_count) + blue + " vowel(s)!")

    print("-" * 30)


def guess_letter():
    global guess

    guess = None
    while guess is None or len(guess) < 1 or len(guess) >= 2 or not guess.isalpha():
        guess = input("Input your guess: ").lower()
        if len(guess) <= 0:
            print("-" * 30)
            print("Don't make an" +yellow+ " empty"+red+" guess!")
            print("-" * 30)
        elif not guess.isalpha():
            print("-" * 30)
            print(red + "Only use letters!")
            print("-" * 30)
        elif len(guess) >= 2:
            print("-" * 30)
            print(red + "Your guess must be exactly " + yellow + "one" +red + " character long!")
            print("-" * 30)
    if guess in already_guessed:
        print("-" * 20)
        print(red + "You already guessed that letter!")
        print("-" * 20)
        guess_letter()
    os.system('cls')


def check_answer():
    global word
    global got_hint1
    global hint
    global lives
    global already_guessed
    global current_word
    global guess
    global fails

    guess_letter()

    indices = []

    if guess in word:
        for idx, value in enumerate(word):
            if value == guess:
                indices.append(idx)
    else:
        lives -= 1
        fails += 1
        if lives == 2:
            get_hint()

        hm.img_print(fails, str(lives))

    for i in indices:
        current_word[i] = guess + " "

    if got_hint1:
        already_guessed.append(hint)
        got_hint1 = False
    already_guessed.append(guess)
    already_guessed.sort()
    guess = ""
    s2 = ""
    for x in current_word:
        s2 = s2 + x
    print(blue + "The current word is: " + yellow + s2)
    print("-" * 50)
    x = []
    for i in current_word:
        s3 = i.replace(" ", "")
        x.append(s3)

    if x == list(word):
        print(blue + "You guessed the word!")
        reset()
        new_game()


def new_game():
    global word
    global already_guessed

    generate_word()

    status = ""
    while status not in ["y", "n"] or status == "!help":
        status = input("Do you want to start a new game? (y/n): ").lower()
        if status == "n":
            exit()
        if status == "!help":
            commands.commands_main()

    empty_word()

    while lives != 0:
        print(green + "Already guessed letters:", already_guessed)

        check_answer()
    print(red + "You lost! The word was: " + yellow + word)
    reset()
    new_game()


def main():
    print(blue + "Welcome to my little Hangman-Game! Made with ♥ by Jackstar#2002!")
    print(blue + "You can see a list of commands, if you type \"!help\" down below")
    print("-" * 75)
    new_game()


if __name__ == "__main__":
    main()
